### ISSUE TITLE
[Question] OnSuccessTry with functions returning "Result"

### ISSUE THREAD

**Issue by rutkowskit** [vkhorikov/CSharpFunctionalExtensions#438]
## Problem Context:
Let's assume we have a function:
```csharp
private static Result DoSomething()
{
    throw new Exception("Simulated exception");        
}
```
Now let's call it using OnSuccessTry like so:
```csharp
var result = Result.Success().OnSuccessTry(DoSomething);
```
The "result" is of type ``` Result<Result> ```.

<b>[Question] Is this behavior intentional ? Or maybe it should return Failure ```Result```  ?</b> <br/>

Right now I am using additional extensions to overcome this behavior. Example for the above scenario:

```csharp
public static Result BindTry(this Result result, Func<Result> func, Func<Exception, string>? errorHandler = null)
{
    return result.IsFailure
        ? Result.Failure(result.Error)
        : Result.Try(() => func(), errorHandler).Bind(r => r);
}
```
I wonder if there should be BindTry and TapTry instead of OnSuccessTry (similar to Bind and Tap instead of OnSuccess), or maybe it is not worth the effort ?

**Comment by vkhorikov** [comment]
We should rename `OnSuccessTry`. Its `OnSuccess` counterparts were already renamed into `Map` or `Bind`; one was left out unintentionally.

I'm not sure which behavior it should have, I haven't used it myself, but having `BindTry` and `TapTry` instead of `OnSuccessTry` is a good idea indeed.

**Comment by rutkowskit** [comment]
Do you mind if I implement the "BindTry" part ?

**Comment by vkhorikov** [comment]
PRs are very much welcome. Thanks.

**Comment by rutkowskit** [comment]
Adding ```BindTry``` for your example code will not work correctly. Even if we assume, that function will return ```Result.Success(JToken.Parse(json))```, the implementation of ```BindTry``` for this situation should be something like this:

```csharp
public static async Task<Result<K, E>> BindTry<T, K, E>(this Task<Result<T, E>> resultTask, Func<T, Result<K>> func, 
            Func<Exception, E> errorHandler)
{
    var result = await resultTask.DefaultAwait();
    if(result.IsFailure)
    {
        return Result.Failure<K, E>(result.Error);
    }
    var funcResult = Result.Try(() => func(result.Value), errorHandler); //Result<Result<K>,E>            
    if(funcResult.IsFailure)
    {
        return Result.Failure<K, E>(funcResult.Error);
    }
    return funcResult.Value.IsSuccess
        ? Result.Success<K,E>(funcResult.Value.Value)
        : Result.Failure<K, E>(default(E)); // what "E" should be returned here ? Returning default value is not very good idea :)
}
```
As you can see, if ```func``` returns failure result, we do not know what Error value should be set in resulting Failure.
In theory it could return ``` Result.Failure<K, E>(errorHandler(new Exception(Result.Error))); ```, but I don't know if it is a good idea, because the Failure is not the result of the exception. 

So I guess what you're looking for  is ```MapTry``` functionality (not yet in the library). <br/>
Example implementation that should work for your code:

```csharp
public static class ResultExtensions
{
    public static async Task<Result<K,E>> MapTry<T,K,E>(this Task<Result<T,E>> taskResult, Func<T, K> func, Func<Exception, E> errorHandler)
    {
        var result = await taskResult;
        return result.MapTry(func,errorHandler);        
    }
    public static Result<K, E> MapTry<T, K, E>(this Result<T, E> result, Func<T, K> func, Func<Exception, E> errorHandler)
    {        
        return result.IsFailure
                ? Result.Failure<K, E>(result.Error)
                : Result.Try(() => func(result.Value), errorHandler).Map(r => r);
    }
}
```

In summary, ```MapTry```, ```TapTry``` and ```CheckTry``` (?) must also be implemented to set ```OnSuccessTry``` as obsolete.

**Comment by rutkowskit** [comment]
Right now I'm setting as ```Obosolete```  the ```OnSuccessTry``` extensions that return Value.
The following ```OnSuccessTry``` extensions remain:
```csharp
public static Result OnSuccessTry(this Result result, Action action, Func<Exception, string> errorHandler = null) { }
public static Result OnSuccessTry<T>(this Result<T> result, Action<T> action, Func<Exception, string> errorHandler = null) {}
public static Result<T, E> OnSuccessTry<T, E>(this Result<T, E> result, Func<T> func, Func<Exception, E> errorHandler) {}
```
I think the ones with ```Action``` on the Right should stay as is, but I don't know what to do with the last one.
The Value from the ```result``` parameter is never used, so it doesn't fit to any "Mapping" extension. 
I guess it was implemented before the ```UnitResult<E>``` (which should be on the Left side in my opinion). 

What do You think ?  Should this last extension method also stay as is, or should it be ```Obsolete``` with message encouraging to explicitly map Result<T,E> to UnitResult<E> to avoid unexpected behavior ?

**Comment by vkhorikov** [comment]
Yeah, let's be consistent here. You are right that `OnSuccessTry` and other `OnSuccessXXX` were introduced before `UnitResult`, hence they don't quite match the other methods semantics-wise.

**Comment by rutkowskit** [comment]
Ok, I've made those changes and create PR. 
If you approve it, then I guess this ```issue``` can be closed as well.

**Comment by vkhorikov** [comment]
Yep, merged the PR and published the changes as v2.34.3. Closing the issue, thanks again.
